src/target/stm32f1_ch32.c

```diff
@@ -0,0 +1,338 @@
+/*
+ * This file is part of the Black Magic Debug project.
+ *
+ * Copyright (C) 2022  Black Sphere Technologies Ltd.
+ * See CH32 Sample code from WCH StdPeriphLib_CH32F1/Examples/FLASH/FLASH_Program
+ *
+ * The CH32 seems to like the EOP bit to be cleared at the end of erase/flash operation
+ * The following code works fine in BMP hosted mode
+ * It does NOT work with a real BMP, only the first 128 bytes block is successfully written
+ *
+ */
+#if PC_HOSTED == 1
+	#define DEBUG_CH DEBUG_INFO
+	#define ERROR_CH DEBUG_WARN
+#else
+	#define DEBUG_CH(...) {} //DEBUG_WARN //(...) {}
+	#define ERROR_CH DEBUG_WARN //DEBUG_WARN
+#endif
+
+
+ static int ch32f1_flash_erase(struct target_flash *f,
+	                            target_addr addr, size_t len);
+ static int ch32f1_flash_write(struct target_flash *f,
+	                            target_addr dest, const void *src, size_t len);
+#define FLASH_MODEKEYR_CH32 (FPEC_BASE+0x24) // Fast mode for CH32F10x
+
+#define FLASH_CR_FLOCK_CH32       (1<<15) // fast unlock
+#define FLASH_CR_FTPG_CH32        (1<<16) // fast page program
+#define FLASH_CR_FTER_CH32        (1<<17) // fast page erase
+#define FLASH_CR_BUF_LOAD_CH32    (1<<18) // Buffer load
+#define FLASH_CR_BUF_RESET_CH32   (1<<19) // Buffer reset
+#define FLASH_SR_EOP              (1<<5)  // End of programming
+#define FLASH_BEGIN_ADDRESS_CH32  0x8000000
+#define FLASH_MAGIC               (FPEC_BASE+0x34)
+
+static volatile uint32_t magic,sr,ct;
```

Comment:
Is it absolutely necessary to carry these as global state? (also, spaces go around operators such as << and +, and after commas)